### ISSUE TITLE
UUID problem

### ISSUE THREAD

**Issue by srrawr11** [kangarko/ChatControl#1216]
Java: 1.16
Version: Paper 1.16.5

I do not know why is showing this to my users.. he is using TLauncher..
![image](https://user-images.githubusercontent.com/73607495/128609286-dba50a52-9a10-48ca-b2df-d81776e97286.png)

and I got this off..
![image](https://user-images.githubusercontent.com/73607495/128609299-209a270b-708c-4653-8269-9278952b22fe.png)

**Comment by ElBananaa** [comment]
> Java: 1.16
> Version: Paper 1.16.5
> 
> I do not know why is showing this to my users.. he is using TLauncher..
> ![image](https://user-images.githubusercontent.com/73607495/128609286-dba50a52-9a10-48ca-b2df-d81776e97286.png)
> 
> and I got this off..
> ![image](https://user-images.githubusercontent.com/73607495/128609299-209a270b-708c-4653-8269-9278952b22fe.png)

Hey!
Please provide us your /chc debug file.

**Comment by srrawr11** [comment]
[debug.zip](https://github.com/kangarko/ChatControl-Red/files/6949822/debug.zip)
Here is it..

Another thing how can I setup the staffchat?
For example in the ChatControlPro I just put used this command an dI could chat with my staffs around the network without using the bungeecord..
![image](https://user-images.githubusercontent.com/73607495/128614087-5b0c154b-55f6-4951-88ec-a671fe1cef54.png)


Another thing, where can I translate this message.. 
![image](https://user-images.githubusercontent.com/73607495/128614079-fe4337eb-59f4-4899-b47c-d0abe8e8e174.png)

**Comment by ElBananaa** [comment]
> [debug.zip](https://github.com/kangarko/ChatControl-Red/files/6949822/debug.zip)
> Here is it..
> 
> Another thing how can I setup the staffchat?
> For example in the ChatControlPro I just put used this command an dI could chat with my staffs around the network without using the bungeecord..
> ![image](https://user-images.githubusercontent.com/73607495/128614087-5b0c154b-55f6-4951-88ec-a671fe1cef54.png)
> 
> Another thing, where can I translate this message..
> ![image](https://user-images.githubusercontent.com/73607495/128614079-fe4337eb-59f4-4899-b47c-d0abe8e8e174.png)

Alright, thanks!
So, it looks like NashornPlus isn't installed.
Please install it first of all : https://mineacademy.org/nashorn/ and let me know if it changes anything.

Regarding the message, it can be translated in `rules/groups.rs`.

**Comment by srrawr11** [comment]
Alright gonna try then, thanks.
Got another question, how can I set up the staffchat?
For example in the ChatControlPro just with /chatcontrol adminchat <message> I could chat with my staffs with around my network.

**Comment by ElBananaa** [comment]
> Alright gonna try then, thanks.
> Got another question, how can I set up the staffchat?
> For example in the ChatControlPro just with /chatcontrol adminchat I could chat with my staffs with around my network.

Hey,

 - To setup a staffchat, you simply need to create a channel.
Let's say the channel is named `staff`, this is how it'll look in your settings file, under Channels.List:
```yaml
  List: 
    default: 
      Format: default-chat
    staff: 
      Format: staff-chat
      Format_Console: '[StaffChat] {player} &8> &r{message}'
```
- Once this is done, make sure your staff members have the permissions `chatcontrol.channel.autojoin.staff.read` and `chatcontrol.channel.join.staff.read`

- Finally, if you want you staff members being able to automatically send a message in the staffchat if their message starts with a `?` then you can create a rule in `rules/chat.rs`. Here's an example that you can past near the top of your file (It must be done at least under `@import global` .
```rs
# -----------------------------------------------------------------------------------------------
# Forwarding messages to other channels.
# -----------------------------------------------------------------------------------------------

# Send messages starting with "?" to staff chat
match ^(\?)(.*)
then command channel send staff $2
then deny
dont verbose
```




If you'd rather use a command, then instead go in `rules/command.rs` and paste the following near the top of your file:
```rs
# -----------------------------------------------------------------------------------------------
# Create a /staffchat command that sends messages to the staff channel.
# -----------------------------------------------------------------------------------------------

# Catch "/staffchat" command and send usage.
match ^\/staffchat$
dont verbose
then warn &cDescription: Send a message to the staffchat.
then warn &cUsage: /staffchat <message>
then deny

# Catch "/staffchat <message>" and redirect to the staff channel.
match ^(\/staffchat) (.*)
dont verbose
then command channel send staff $2
then deny
```
Of course, you're free to edit the commands, to use both methods, or to name the channel however you want, these are only used as examples.


For more informations, I highly recommend taking a look at our wiki on which you can find all informations to do so. You can also take a look at all the default files since they also include some examples and a lot of comments to explain how these examples work.
Here are some useful wiki links:
https://github.com/kangarko/ChatControl-Red/wiki/Permissions
https://github.com/kangarko/ChatControl-Red/wiki/Channels
https://github.com/kangarko/ChatControl-Red/wiki/Rules
https://github.com/kangarko/ChatControl-Red/wiki/Custom-Commands (This page seems a bit outdated as for now, since I only noticed it right now, I'd suggest waiting a bit and looking at the examples that are already in the plugin files instead).

**Comment by kangarko** [comment]
Thanks to Banana for helping out! I have reviewed your files but could not reproduce the `<chat=??>` problem (https://i.imgur.com/vRLPV6H.png), perhaps it was related to an outdated ProtocolLib version, and ensure there is not errors in the console during startup etc., if yes, just post them.

**Comment by srrawr11** [comment]
Sorry for the late answers, I appreciate how you can trying help us guys!
When I activate the channel section, my users cant chat, it says, something about /ch join.. but is it posible to join automatically? and how?

In the ChatControlPro was easy to set up this part of staffchat.. :/

Second thing, It crashes the whole server..
[error.log](https://github.com/kangarko/ChatControl-Red/files/6995080/error.log)

**Comment by ElBananaa** [comment]
Hey!
The issue you're having is pretty weird.
Please update to the latest version of the plugin.
Also, in your error log file, it says your server is under Bukkit 1.8.8/java 8, but your issue says java 16 & paper 1.16.5, why?
Try to update this server to spigot/paper 1.8 instead of bukkit.
Finally, to automatically join channels, please refer to our wiki: https://github.com/kangarko/ChatControl-Red/wiki/Channels & https://github.com/kangarko/ChatControl-Red/wiki/Channels#auto-join

**Comment by srrawr11** [comment]
Thanks I could set up, I appreciate your time and help.
One last thing, do you know how what is the permission that onlt my staffs can read the staff chat?
Coz for example if I put /sc Hi.. they can't read it..

And do I need activate mysql for talk server to server? Like ... If I am in Survival and I wrote /sc Hi and my Jail staffs can read that?

**Comment by kangarko** [comment]
If you have BungeeCord, then yes, you need MySQL.

You will need to use our Channels and create an admin channel for your staff. For permissions and examples, please see our Wiki on GitHub and the Channels section.

**Comment by srrawr11** [comment]
Yes, thanks, I solved it!
Thanks to you guys! :)

Keep safe!

**Comment by ElBananaa** [comment]
> Yes, thanks, I solved it!
> Thanks to you guys! :)
> 
> Keep safe!

We're glad you finally solved your issues! 😄
Stay safe aw!

If you need anything else, feel free to open a new issue, we'd be more than happy to help 😉

**Comment by srrawr11** [comment]
Thanks :)

I need to know why I can not send private msg to my users..
For example is saying this..

![image](https://user-images.githubusercontent.com/73607495/130358262-ff6d0e69-2046-481a-928e-11595efd604e.png)

Where he is already connected.

**Comment by ElBananaa** [comment]
> Thanks :)
> 
> I need to know why I can not send private msg to my users..
> For example is saying this..
> 
> ![image](https://user-images.githubusercontent.com/73607495/130358262-ff6d0e69-2046-481a-928e-11595efd604e.png)
> 
> Where he is already connected.

Hey,
Please update to the latest version of the plugin. If you have a network, please update the plugin on all servers, and update bungeecontrol-red as well.

**Comment by kangarko** [comment]
There is going to be another update coming up tonight (althought marked as beta) you can test out

**Comment by srrawr11** [comment]
That's great, Im gonna try out then.
One thing, I would like to add this for my users and vips ranks. I did somehting like this, I read the wiki but still doesnt work.
![image](https://user-images.githubusercontent.com/73607495/130682918-285c48f0-f7d1-4325-a5a2-30acacb943c8.png)
 MC is one of my VIP rank, but the users who got the rank it doesnt shows like I configurated.

**Comment by ElBananaa** [comment]
> That's great, Im gonna try out then.
> One thing, I would like to add this for my users and vips ranks. I did somehting like this, I read the wiki but still doesnt work.
> ![image](https://user-images.githubusercontent.com/73607495/130682918-285c48f0-f7d1-4325-a5a2-30acacb943c8.png)
> MC is one of my VIP rank, but the users who got the rank it doesnt shows like I configurated.

Hey!
This isn't how the `group whatever-name` works. Refer to https://github.com/kangarko/ChatControl-Red/wiki/Rules#groups

You need to use either the `require perm` variable (https://github.com/kangarko/ChatControl-Red/wiki/Rules#require-perm-permission-message) or a custom condition like the example below to do what you want:
```rs
group default-join-message
require sender script "{player}" == "Bananaa"
message:
- &6[&a+&6] &d{player}
```
That's up to you to replace the `{player}` placeholder with the rank placeholder, and the `Bananaa` with the rank name.

**Comment by srrawr11** [comment]
I have a question, how can I give a chat color auto to my ranks?
Like &9&lVIP I woul like to give them the chat color ''&9Hi''

**Comment by ElBananaa** [comment]
> I have a question, how can I give a chat color auto to my ranks?
> Like &9&lVIP I woul like to give them the chat color ''&9Hi''

Hey,
Apologize for the delay.

There are multiple ways to do so. but in your case, I'd suggest using our format system.
Basically, in the default `format/chat.yml` file, there is a `message:` part (At the very bottom of the file), which looks like this:
```yaml
  # And finally, show the message the player has typed out, prepending it with player's
  # suffix from your permission plugin. If player has custom '/chc color', we append it
  # after the suffix. Even if you remove the suffix, custom colors will still be appended. 
  message:
    Message: '&7: {player_suffix}{message}'
```
Simply edit this part using our condition system (more informations on ou wiki page https://github.com/kangarko/ChatControl-Red/wiki/Formats). Here's an example that you can use and edit:
```yaml
  default-message:
    Message: '&7: &r{message}'
    Sender_Condition: '''{player_group}'' == ''default'''

  vip-message:
    Message: '&7: &9{message}'
    Sender_Condition: '''{player_group}'' == ''VIP'''

  moderator-message:
    Message: '&7: &c{message}'
    Sender_Condition: '''{player_group}'' == ''moderator'''
```

**Comment by srrawr11** [comment]
Hi, thanks to answer me, that quite complicate, I do not understand.. :/ and the quit and join for VIP ranks as well. I have tried all.. but nothing..

Here is my config..
Chat: https://pastebin.com/pPY4Tj47
Join/Quit: https://pastebin.com/AmcLLw6Z ||| https://pastebin.com/hBLdCBRS

**Comment by ElBananaa** [comment]
> Hi, thanks to answer me, that quite complicate, I do not understand.. :/ and the quit and join for VIP ranks as well. I have tried all.. but nothing..
> 
> Here is my config..
> Chat: https://pastebin.com/pPY4Tj47
> Join/Quit: https://pastebin.com/AmcLLw6Z ||| https://pastebin.com/hBLdCBRS

Hey!
I don't understand what's the issue with your chat file? Everything inside seems good to me.

Regarding your Join/Quit messages, it doesn't work because you used `require sender script "{player}" == "MC"` and `require sender script "{player}" == "MC+"`.

Basically, this variable means that the message is only sent if the player who joins is named `MC` or `MC+`.
To check for a group, you need to use the placeholder that returns the group name instead of `{player}` which returns the player name.
So using `require sender script "{player_group}" == "MC"` & `require sender script "{player_group}" == "MC+"` instead of what you're using right now should work.

**Comment by srrawr11** [comment]
yes the chat one I one to put like each group with a different chat colour.
For example MC is one of my VIP rank and the color is RED, but I do not know how to configurate.

and the join/quit is not working.. I'm using LuckPerms for the ranks.
![image](https://user-images.githubusercontent.com/73607495/132877297-24b726e1-6d10-4e2b-a16a-bef234505510.png)

**Comment by ElBananaa** [comment]
> yes the chat one I one to put like each group with a different chat colour.
> For example MC is one of my VIP rank and the color is RED, but I do not know how to configurate.
> 
> and the join/quit is not working.. I'm using LuckPerms for the ranks.
> ![image](https://user-images.githubusercontent.com/73607495/132877297-24b726e1-6d10-4e2b-a16a-bef234505510.png)

Alright, then I guess the chat thing is working, great!

The join/quit messages aren't working because, according to the debug file you provided us early august, they're not enabled.
Enable them in the settings file under `Messages.Apply_On` (Like on the screenshot below), and it should work.
![image](https://user-images.githubusercontent.com/53318613/132880521-942e69ba-1308-4b1c-97a3-4c7a7d0fb273.png)

**Comment by srrawr11** [comment]
Already got them on the config activated.. but same, it doesn't work for my MC ranks users..
![image](https://user-images.githubusercontent.com/73607495/132894152-40b390e9-b7bb-40bd-be23-2a376fa34e36.png)
![image](https://user-images.githubusercontent.com/73607495/132894192-b8ade282-d687-414b-8494-b1f8c510357e.png)
In the second pic as you can see.. I have told him to go to the /lobby and join the server.. but it's same.

**Comment by ElBananaa** [comment]
Ok, I think I found the issue:
Make sure to place the custom join/quit messages of each ranks above the default join message in your files.

**Comment by srrawr11** [comment]
Same, I have done what did you just said..
![image](https://user-images.githubusercontent.com/73607495/132895122-c4161e4a-06bd-4c81-a897-75976cff8aea.png)

![image](https://user-images.githubusercontent.com/73607495/132895142-788b909a-1397-47cb-839c-e4954314a4e3.png)

**Comment by ElBananaa** [comment]
> Same, I have done what did you just said..
> ![image](https://user-images.githubusercontent.com/73607495/132895122-c4161e4a-06bd-4c81-a897-75976cff8aea.png)
> 
> ![image](https://user-images.githubusercontent.com/73607495/132895142-788b909a-1397-47cb-839c-e4954314a4e3.png)

Then it means the group name isn't exactly `MC` or `MC+`.
Please run `/chc info variable <player> {player_group}` (Of course, replace `<player>` with the player name. So, fo me it would be `/chc info variable ElBananaa {player_group}`), and provide us a screenshot of what it returns.

**Comment by srrawr11** [comment]
But the MC o MC+ is not a username.. its just a RANK created with LUCKPERMS, is one of my RANKS that I sell to the shop as 'VIP' rank. I do not know if I could explained properly.

**Comment by ElBananaa** [comment]
> But the MC o MC+ is not a username.. its just a RANK created with LUCKPERMS, is one of my RANKS that I sell to the shop as 'VIP' rank. I do not know if I could explained properly.

Yes, i know that, that's why I asked you to use the command above. Just use it and send ud the screenshots.

**Comment by srrawr11** [comment]
I executed this command.. /chc info variable jeremixd {player_group} .. is one of the users who bought MC+
![image](https://user-images.githubusercontent.com/73607495/133001209-b7c99601-3048-4004-bc41-d65de5e79343.png)

**Comment by ElBananaa** [comment]
> I executed this command.. /chc info variable jeremixd {player_group} .. is one of the users who bought MC+
> ![image](https://user-images.githubusercontent.com/73607495/133001209-b7c99601-3048-4004-bc41-d65de5e79343.png)

Alright, please try to replace MC and MC+ with mc and mc+ (lowercased) in your rule, and see it it helps

**Comment by srrawr11** [comment]
Sorry for the late answer, I appreciate that you are trying to help me :) seriously!
But is still not working.. I do not know what I'm doing wrong..

![image](https://user-images.githubusercontent.com/73607495/133940464-085b6530-57ac-4862-bd14-ec26fdf5e7d9.png)

**Comment by ElBananaa** [comment]
Hey, thanks, and no worries ^^
Alright, so keep the changes you made for the ranks (keep the lowercased names), but instead of using the `{player_group}` placeholder, install luckperms PlaceholderAPI expansion by running : `/papi ecloud download LuckPerms` and `/papi reload`, then replace `{player_group}` with `{luckperms_primary_group_name}`.
If it still doesn't work, try to run `/papi parse <player> %luckperms_primary_group_name%` (don't forget to replace `<player>` with the name of an online player who either has the mc or mc+ rank), and send us the screenshot.

**Comment by srrawr11** [comment]
Yes, now its working, just was the placeholderAPI.. I have changed {player_group} to {luckperms_primary_group_name}, thanks :)

One thing more, I still didnt fixed the chat colour for the ranks..

**Comment by ElBananaa** [comment]
> Yes, now its working, just was the placeholderAPI.. I have changed {player_group} to {luckperms_primary_group_name}, thanks :)
> 
> One thing more, I still didnt fixed the chat colour for the ranks..

Well, in your format file, simply edit the `{message}` part to work based on ranks.
Here's an example:
```yaml
  default-message:
    Message: '&7: &r{message}'
    Sender_Condition: '''{luckperms_primary_group_name}'' == ''default'''

  vip-message:
    Message: '&7: &9{message}'
    Sender_Condition: '''{luckperms_primary_group_name}'' == ''VIP'''

  moderator-message:
    Message: '&7: &c{message}'
    Sender_Condition: '''{luckperms_primary_group_name}'' == ''moderator'''
```
Simply edit this to make it suits you the most.

**Comment by srrawr11** [comment]
Ahh done, I got it, thanks mate! For you time and dedication! I appreciated :)

**Comment by ElBananaa** [comment]
> Ahh done, I got it, thanks mate! For you time and dedication! I appreciated :)

Glad it worked!
You're welcome. If you need anything else, feel free to open a new issue 😉

**Comment by srrawr11** [comment]
Hey again!
I would like to know how can I activate the formatting colours? Example..
&6&lHi, how are you?

Coz when I do that, it doesn't work.

**Comment by ElBananaa** [comment]
Hey,
In your settings file, look for `Color.Apply_On` (starting at line 575 in the default settings file), and make sure colors are enabled
Here's how it looks like:
```yaml
# -------------------------------------------------------------------------------------------------
# Represents settings for letting players use colors in their chat or select a custom color.
#
# You must give players the following permissions for them to be able to use colors:
# - chatcontrol.color.{color} (replace color with color name such as "red)
# - chatcontrol.hexcolor.{color} (replace color with the hex code such as "ccdd44")
# - chatcontrol.guicolor.{color} (replace color with color names, used in /chc color menu)
# - chatcontrol.use.color.{apply_on} (replace apply_on with options below) - Let players write &
#                                  colors and hexcodes to places in Apply_On below. true by default
# -------------------------------------------------------------------------------------------------
Colors:
 
  # Where should we let players to use & or RGB color codes?
  # Options: chat, me, prefix, nick, suffix, private_message
  Apply_On: [chat, me, prefix, nick, suffix, private_message]
```

Also keep in mind that players need to have specific permissions to use colors.
Everything is explained in the comments above this part.

**Comment by srrawr11** [comment]
Ah yeah, but for example how can I just activate the codes &abcdef and &1234567890?
I have to go one by one like..
chatcontrol.color.red
chatcontrol.color.yellow
ando so on.. ?

**Comment by ElBananaa** [comment]
> Ah yeah, but for example how can I just activate the codes &abcdef and &1234567890?
> I have to go one by one like..
> chatcontrol.color.red
> chatcontrol.color.yellow
> ando so on.. ?

I'm not sure about it, but giving them `chatcontrol.color.*` and `chatcontrol.hexcolor.*` should work.
Of course, don't forget to also give them the permission to use colors in specific parts of the chat using `chatcontrol.use.color.{apply_on}` (Afaik, this one doesn't support `chatcontrol.use.color.*` so you'll have to give users `chatcontrol.use.color.chat`, `chatcontrol.use.color.me`, `chatcontrol.use.color.private_message`... to use colors everywhere).

**Comment by srrawr11** [comment]
Alright, solved, thanks again :)
Have a nice day and keep safe!

**Comment by srrawr11** [comment]
Hi, It just a quick question..
Some users does this type of spam in my server..
![image](https://user-images.githubusercontent.com/73607495/140642770-1fa48add-408b-4ee1-bde4-32ec369f55d6.png)

The problem is.. I have disable the section of that because when If I do /pay <nick> 1000000000 .. something like that.. it takes as a spam aswell.. so how can I fix this?

Thanks again and hope you are fine! :)

**Comment by ElBananaa** [comment]
Hey,
Simply add the `/pay` command to the whitelisted command section of the anti_spam config (look in the settings file, you'll find it)

**Comment by srrawr11** [comment]
Thank you so much again :)
